### PR TITLE
{2023.06}[GCCcore/12.2.0][GCCcore/12.3.0] BioPerl v1.7.8

### PR DESCRIPTION
**Pull request by TopRichard**
BioPerl build process tries to copy `perlmodule.py` once again after sanity check which results in  :

```
Failed to copy file /cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/intel/haswell/software/EasyBuild/4.9.2/lib/python3.11/site-packages/easybuild/easyblocks/generic/perlmodule.py to .../reprod_20240802080341_2187/e
asyblocks/perlmodule.py: [Errno 13] Permission') (at easybuild/main.py:178 in build_and_install_software)

```

BioPerl hook adds user write permission after sanity check to allow copying the easyblock file.

Missing packages:

```
BioPerl/1.7.8-GCCcore-12.2.0
BioPerl/1.7.8-GCCcore-12.3.0
DB/18.1.40-GCCcore-12.3.0
DB_File/1.859-GCCcore-12.3.0
groff/1.22.4-GCCcore-12.3.0
Perl-bundle-CPAN/5.36.1-GCCcore-12.3.0
XML-LibXML/2.0208-GCCcore-12.2.0
XML-LibXML/2.0209-GCCcore-12.3.0
```

### CODE REVIEW

**Comment by TopRichard** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/generic
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/haswell
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/skylake_avx512
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen2
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen3
bot: build repo:eessi.io-2023.06-software arch:aarch64/generic
bot: build repo:eessi.io-2023.06-software arch:aarch64/neoverse_n1
bot: build repo:eessi.io-2023.06-software arch:aarch64/neoverse_v1

File: eb_hooks.py [hunk 1]
```diff
@@ -647,6 +647,16 @@ def post_sanitycheck_hook(self, *args, **kwargs):
         POST_SANITYCHECK_HOOKS[self.name](self, *args, **kwargs)
 
 
+def post_sanitycheck_bioperl(self, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (ocaisa) [reply eb_hooks.py:650]: If this is really an issue with the permissions of the easyblock then it needs generic fixing in EasyBuild (in framework, so it doesn't happen again)

**Comment by TopRichard** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/generic

**Comment by TopRichard** [comment]
After removing the hook, the build fails :
```
== permissions...
== ... (took < 1 sec)
== packaging...
== ... (took < 1 sec)
== FAILED: Installation ended unsuccessfully (build directory: /tmp/bot/easybuild/build/BioPerl/1.7.8/GCCcore-12.2.0): build failed (first 300 chars): Failed to copy file /cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/generic/software/EasyBuild/4.9.2/lib/python3.11/site-packages/easybuild/easyblocks/generic/perlmodule.py to /tmp/eb-fftevtnk/eb-ak3d7syr/reprod_20240805124945_346/easyblocks/perlmodule.py: [Errno 13] Permission denie (took 1 min 36 secs)
== Results of the build can be found in the log file(s) /tmp/eb-fftevtnk/eb-ak3d7syr/easybuild-BioPerl-1.7.8-20240805.124945.bGmIN.log
0:02:13  1 out of 2 easyconfigs done: XML-LibXML/2.0208-GCCcore-12.2.0 (OK)ERROR: Build of /cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/generic/software/EasyBuild/4.9.2/easybuild/easyconfigs/b/BioPerl/BioPerl-1.7.8-GCCcore-12.2.0.eb failed (err: 'build failed (first 300 chars): Failed to copy file /cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/generic/software/EasyBuild/4.9.2/lib/python3.11/site-packages/easybuild/easyblocks/generic/perlmodule.py to /tmp/eb-fftevtnk/eb-ak3d7syr/reprod_20240805124945_346/easyblocks/perlmodule.py: [Errno 13] Permission denie')
Last EasyBuild log file copied from /tmp/eb-fftevtnk/easybuild-48ln20pb.log to /project/60006/SHARED/jobs/2024.08/pr_658/event_50dd1c80-5328-11ef-93bd-58a74643a660/run_000/linux_x86_64_generic/eessi.io-2023.06-software
EasyBuild log file /tmp/eb-fftevtnk/easybuild-48ln20pb.log copied to /project/def-users/SHARED/build-logs/jobs/15702/easybuild-48ln20pb.log (with context appended)
Using /project/60006/SHARED/jobs/2024.08/pr_658/event_50dd1c80-5328-11ef-93bd-58a74643a660/run_000/linux_x86_64_generic/eessi.io-2023.06-software/658.diff to give create exceptions for PR filtering of easystack
```

**Comment by ocaisa** [comment]
So this seems like the file exists already for whatever reason, and an overwriting copy fails. This sounds like an EasyBuild framework problem, it should
* check if the file exists
  * if it does, check if the contents are the same. If they are don't bother trying to copy, if they are not error out
  * if it does not, make the copy

**Comment by ocaisa** [comment]
I think you can probably cheat your way around it by using `--include-easyblocks-from-commit=89ea2e4737d792055de828e18f38bccee26b76cf` (latest commit to easyblock repo), that will prefer the easyblocks from that repo which won't have these restrictive permissions.

An issue should still be opened in the EasyBuild framework

**Comment by TopRichard** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/generic
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/haswell
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/skylake_avx512
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen2
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen3
bot: build repo:eessi.io-2023.06-software arch:aarch64/generic
bot: build repo:eessi.io-2023.06-software arch:aarch64/neoverse_n1
bot: build repo:eessi.io-2023.06-software arch:aarch64/neoverse_v1

**Comment by boegel** [comment]
> I think you can probably cheat your way around it by using `--include-easyblocks-from-commit=89ea2e4737d792055de828e18f38bccee26b76cf` (latest commit to easyblock repo), that will prefer the easyblocks from that repo which won't have these restrictive permissions.
> 
> An issue should still be opened in the EasyBuild framework

I'm not a big fan of using a workaround like this...

I'm also puzzled: why does this only occur with `BioPerl`, which is a pretty standard easyconfig using the `PerlModule` easyblock? Is it because the top-level easyblock being used is `PerlModule`, and then the extensions are also installed with that same easyblock (which is kind of weird)?

**Comment by ocaisa** [comment]
I've left a comment in https://github.com/easybuilders/easybuild-framework/issues/4602#issuecomment-2271082442 about (at least one place) where I think this could be resolved in framework, but that won't help us here as that would need a new release for us to leverage.

**Comment by TopRichard** [comment]
> I think you can probably cheat your way around it by using `--include-easyblocks-from-commit=89ea2e4737d792055de828e18f38bccee26b76cf` (latest commit to easyblock repo), that will prefer the easyblocks from that repo which won't have these restrictive permissions.
> 
> An issue should still be opened in the EasyBuild framework

The workaround doesn't work

**Comment by ocaisa** [comment]
Ok, I found the problem, the easyconfigs are using the `PerlModule` easyblock and so are the extensions: https://github.com/easybuilders/easybuild-easyconfigs/blob/develop/easybuild/easyconfigs/b/BioPerl/BioPerl-1.7.8-GCCcore-12.2.0.eb
That's a departure from previous easyconfigs

**Comment by ocaisa** [comment]
Fix in https://github.com/easybuilders/easybuild-easyconfigs/pull/21136

File: easystacks/software.eessi.io/2023.06/eessi-2023.06-eb-4.9.2-2022b.yml [hunk 2]
```diff
@@ -1,2 +1,6 @@
 easyconfigs:
   - BLAST+-2.14.0-gompi-2022b.eb
+  - BioPerl-1.7.8-GCCcore-12.2.0.eb:
+      options:
+        # see https://github.com/easybuilders/easybuild-framework/issues/4602 
+        include-easyblocks-from-commit: 89ea2e4737d792055de828e18f38bccee26b76cf
```
Comment (ocaisa) [inline easystacks/software.eessi.io/2023.06/eessi-2023.06-eb-4.9.2-2022b.yml:6]:
```suggestion
        # see https://github.com/easybuilders/easybuild-easyconfigs/pull/21136
        from-commit: d8076ebaf8cb915762adebf88d385cc672b350dc 
```

File: easystacks/software.eessi.io/2023.06/eessi-2023.06-eb-4.9.2-2023a.yml [hunk 3]
```diff
@@ -36,3 +36,7 @@ easyconfigs:
         include-easyblocks-from-commit: c8256a36e7062bc09f5ce30552a9de9827054c9e
         # https://github.com/easybuilders/easybuild-easyconfigs/pull/20841
         from-commit: f0e91e6e430ebf902f7788ebb47f0203dee60649
+  - BioPerl-1.7.8-GCCcore-12.3.0.eb:
+      options:
+        # see https://github.com/easybuilders/easybuild-framework/issues/4602 
+        include-easyblocks-from-commit: 89ea2e4737d792055de828e18f38bccee26b76cf
```
Comment (ocaisa) [inline easystacks/software.eessi.io/2023.06/eessi-2023.06-eb-4.9.2-2023a.yml:42]:
```suggestion
        # see https://github.com/easybuilders/easybuild-easyconfigs/pull/21136
        from-commit: d8076ebaf8cb915762adebf88d385cc672b350dc
```

**Review by ocaisa — CHANGES_REQUESTED** [review]

**Review by ocaisa — APPROVED** [review]
LGTM

**Pull request merged**